spec.md

```diff
@@ -547,6 +547,19 @@ Analog<32> ; 32-bit analog type
 Analog     ; analog type with inferred width
 ```
 
+### Enumeration Types
+
+Enumerations are structural disjoint union types.  An enumeration has a number
+of variants, each with a type.  The different variants are specified with tags.
+A variant may optionally omit the data type, in which case it is implicitly
+defined to be `UInt<0>`. The variant types of an enumeration must all be
+passive and cannot contain analog types.
+
+``` firrtl
+[#a, #b, #c]
+[#some: UInt<8>, #none]
+```
```

Comment:
It's fine if we don't allow defined enumeration encodings initially, but I think it's going to be a necessary extension so hopefully we have an idea of how that should look in the spec.

Reply:
Yeah, I am guessing the two things we will want are to specify the width of the tags and the values of the tags.  I think specifying the values of tags should be easy (e.g. `[#a = 1, #b = 2, #c = 4]`), but I'm not sure how to specify the overall width.  For any more complicated encoding, it would probably be best to write a chisel decode helper function to convert from a UInt to the enum type.

Reply:
You'll need to be able to specify width because users may want a one-hot encoding (as you showed in your example 🙂).

Per more complicated encodings, yeah I'm not sure if/how we could represent encodings that are more complicated than a 1-1 relationship between the tag and some numerical value or bitvector. It's an interesting question but probably beyond the scope of this current work.

Reply:
I'm starting to wonder if specifying the width explicitly is actually helpful, I can't really figure out how it would be observable. Right now the default width is just the minimum required to hold the largest tag value.

We don't have the ability to "unsafe bitcast" a random `UInt` to an enum type, and even if we did you could just extract the right bits for the tag and cast that (and if there is data, concat tag bits with the data bits):
```
input i : UInt<4>
output o : [#a = 0, #b = 1]
o <= asEnum<[#a = 0, #b = 1]>(head(tail(i, 1), 1)
```

FWIW right now you have to manually write some kind of decoder, which I hope should usually be quite cheap in HW.
```
input i : UInt<4>
output o : [#a = 0, #b = 1]
node n = head(tail(i, 1), 1)
when (n == 0):
  o <= #a : [#a = 0, #b = 1]
else when (n == 1)
  o <= #b : [#a = 0, #b = 1]
else:
  o is invalid
```

Reply:
I do think type-unsafe casts are going to be wanted eventually.  You want to be able to specify an enum which matches a standards-defined bit-layout (within reason) and not have to write a reencoder.